### PR TITLE
[feature] add pytorch profiling

### PR DESCRIPTION
**Pull request by winglian**
Screenshot 2024-12-12 at 8 48 59 PM

### CODE REVIEW

**Review by djsaunde — APPROVED** [review]
LGTM, but can we add inline comments and / or some documentation of this config option?

File: src/axolotl/utils/callbacks/profiler.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+class PytorchProfilerCallback(TrainerCallback):
+    """
+    PyTorch Profiler callback to create snapshots of GPU memory usage at specified steps.
+    """
+
+    def __init__(self, steps_to_profile: int = 5):
+        self.steps_to_profile = steps_to_profile
+        if self.steps_to_profile:
+            torch.cuda.memory._record_memory_history(  # pylint: disable=protected-access
```
Comment (NanoCode012) [inline src/axolotl/utils/callbacks/profiler.py:24]: Is there any cost to recording all memory here (ie increased ram usage etc)?

File: src/axolotl/utils/callbacks/profiler.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    def on_step_end(  # pylint: disable=unused-argument
+        self,
+        args: TrainingArguments,  # pylint: disable=unused-argument
+        state: TrainerState,
+        control: TrainerControl,  # pylint: disable=unused-argument
+        **kwargs,  # pylint: disable=unused-argument
+    ):
+        if state.global_step == self.steps_to_profile:
+            snapshot = torch.cuda.memory._snapshot()  # pylint: disable=protected-access
+            with open(Path(args.output_dir) / "snapshot.pickle", "wb") as fout:
```
Comment (NanoCode012) [inline src/axolotl/utils/callbacks/profiler.py:37]: Can we maybe output to a non-pickle file for safety reasons (shared to others , tampered, etc)? How would a person also read this file and display it?

Comment (winglian) [reply src/axolotl/utils/callbacks/profiler.py:37]: I don't believe there are any other options that the visualizer supports as input. I added docs but forgot to push them earlier.

**Pull request merged**